### ISSUE TITLE
SAMap and Docker

### ISSUE THREAD

**Issue by avianalter** [atarashansky/SAMap#115]
A SAMap Docker image doesn't really exist, so I went ahead and made one. It is missing map_genes.sh and the jupyter notebook, but otherwise has all dependencies and extra stuff that are required for going from raw 10X data to the SAMap sample jupyter notebook. This includes required python libraries and stuff, but also includes things that I found useful, such as R/Seurat and R/sceasy, which I used to go from cellranger output to a '$whatever.h5ad' file needed for SAMap or gffread, which I used to go from reference genome to usable transcriptome for map_genes.sh (c.f. #114), etc. 

https://hub.docker.com/r/avianalter/samap

I'm adding my Dockerfile below rather than uploading it on github because I know that it will be broken by the time you see this, as the keys for updating keys for the ubuntu certificate or the keys for the cellranger downloads will likely be different by then. Feel free to edit for your own purposes if building from scratch. It should be noted that the Docker image is likely bloated as I didn't perform any kind of cleanups and there are some inefficiencies, but the image on dockerhub does work correctly. 

Dockerfile
--------------------------------------
> FROM ubuntu:20.04 AS builder
> ENV TZ=America/New_York
> RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
> RUN apt-get update
> RUN apt-get install -y gnupg ca-certificates libssl-dev curl libcurl4-openssl-dev
> RUN echo 'deb https://cloud.r-project.org/bin/linux/ubuntu focal-cran40/' >> /etc/apt/sources.list
> RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 51716619E084DAB9
> RUN apt-get update
> RUN apt-get install -y build-essential wget git r-base r-base-dev libcairo2-dev libxt-dev
> RUN Rscript -e 'install.packages(c("remotes", "devtools", "SeuratObject"), repos="https://cloud.r-project.org")'
> RUN Rscript -e 'install.packages(c("parallelly", "listenv", "globals", "dotCall64", "future"), repos="https://cloud.r-project.org")'
> RUN Rscript -e 'install.packages(c("RcppEigen", "generics", "spam", "progressr", "future.apply", "sp"), repos="https://cloud.r-project.org")'
> RUN Rscript -e 'install.packages(c("Seurat"), repos="https://cloud.r-project.org")'
> RUN Rscript -e '{if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")};BiocManager::install(c("LoomExperiment", "SingleCellExperiment"))'
> RUN Rscript -e 'install.packages(c("reticulate"), repos="https://cloud.r-project.org")'
> RUN Rscript -e 'remotes::install_github("cellgeni/sceasy")'
> RUN apt-get install -y curl
> #this command will need to be adjusted based on licensing
> RUN mkdir /apps && cd /apps && curl -o cellranger-7.1.0.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-7.1.0.tar.gz?Expires=1683348760&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly9jZi4xMHhnZW5vbWljcy5jb20vcmVsZWFzZXMvY2VsbC1leHAvY2VsbHJhbmdlci03LjEuMC50YXIuZ3oiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2ODMzNDg3NjB9fX1dfQ__&Signature=oQ8mvdJH1jiE2Z4z7Bq~8XSqZgkIn9cimV7OybBO6zHYKwwEqU4uOQ2WPDn0O0cgXMf-d7NfxnB4KajdAZHXPIpJCreXYBvQENhMJ3JgS63vaL~hpZN8zR4xt3pYRmytaKcjoLaOoxZFg4F8fJmACOLrg8E4ClEeEwNZNkNQFDeYDrw5yy7D8YECH2B9FTxIqAOpGk~7fqIgQQ4PDyQQZyoEyMoRCk4o7t6FC2YPYwC-kuWwATZJQCabN7rtdhI--i7ASlnxFyS4fgIe8K7DSoxnq2JAYJ2PCT5rS88MIfumFRBukOUjiglwbSTU3BQgYiiunQ4Q4a72GP7-WP6WsA__&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"
> RUN cd /apps && tar -xzvf *.tar.gz
> ENV PATH=$PATH:/apps/cellranger-7.1.0
> ENV HNSWLIB_NO_NATIVE=1  
> RUN apt-get install -y python3.8 python3-dev python-dev
> RUN wget https://bootstrap.pypa.io/get-pip.py && python3 get-pip.py
> RUN pip3 install samap
> ARG ncbi_blast_version='2.9.0'
> RUN cd /apps && wget "ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/${ncbi_blast_version}/ncbi-blast-${ncbi_blast_version}+-x64-linux.tar.gz" 
> RUN cd /apps && tar -xzvf "ncbi-blast-${ncbi_blast_version}+-x64-linux.tar.gz" -C /usr/bin/ --strip-components=2 "ncbi-blast-${ncbi_blast_version}+/bin/"
> RUN Rscript -e 'install.packages(c("dplyr", "patchwork"), repos="https://cloud.r-project.org")'
> RUN wget https://repo.anaconda.com/miniconda/Miniconda3-py38_23.3.1-0-Linux-x86_64.sh
> RUN bash Miniconda3-py38_23.3.1-0-Linux-x86_64.sh -b
> ENV PATH=$PATH:/root/miniconda3/bin/
> RUN conda init bash
> RUN . /root/.bashrc
> RUN conda install -c conda-forge -y loompy 
> RUN cp /root/miniconda3/lib/libstdc++.so.6.0.29 /usr/lib/x86_64-linux-gnu/
> RUN rm /usr/lib/x86_64-linux-gnu/libstdc++.so.6
> RUN ln -s /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.29 /usr/lib/x86_64-linux-gnu/libstdc++.so.6
> RUN conda install -c conda-forge -y anndata
> RUN pip3 install jupyterlab notebook
> RUN . /root/.bashrc && pip3 install samap
> RUN conda install -c bioconda gffread

**Comment by atarashansky** [comment]
Hi @avianalter ! That docker image is quite old - I removed the Dockerfile since I didn't have the bandwidth to maintain it. Did you update the version of SAMap to the latest in the docker image?

**Comment by avianalter** [comment]
@atarashansky very understandable re: the Dockerfile. 

Originally, I installed everything from scratch into a new Docker image with a Dockerfile I wrote. Got everything to work, but the SAMGUI functionality was broken. Wasted entirely too much time in dependency hell trying to get plotly to play nice with Jupyter so I decided to just try your Docker image instead and got everything to work. 

That being said, I just upgraded the SAMap version in the image manually and then pushed it to Dockerhub just now. I'm a bit too scared to use setuptools because I'm worried it will break a bunch of dependencies, so I literally just copy/pasted the newest *.py files into the miniconda installation. Ran your vignette in Jupyter and the import, SAM initialization, SAMAP, and SAMGUI functionalities worked fine. 

Thanks for the software!
